src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleLogFormatter.cs

```diff
@@ -0,0 +1,168 @@
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Extensions.Logging.Console
+{
+    internal class JsonConsoleLogFormatter : ConsoleFormatter, IDisposable
+    {
+        private IDisposable _optionsReloadToken;
+
+        public JsonConsoleLogFormatter(IOptionsMonitor<JsonConsoleFormatterOptions> options)
+            : base (ConsoleFormatterNames.Json)
+        {
+            FormatterOptions = options.CurrentValue;
+            ReloadLoggerOptions(options.CurrentValue);
+            _optionsReloadToken = options.OnChange(ReloadLoggerOptions);
+        }
+
+        public override void Write<TState>(in LogEntry<TState> logEntry, IExternalScopeProvider scopeProvider, TextWriter textWriter)
+        {
+            LogLevel logLevel = logEntry.LogLevel;
+            string category = logEntry.Category;
+            int eventId = logEntry.EventId.Id;
+            string message = logEntry.Formatter(logEntry.State, logEntry.Exception);
+            Exception exception = logEntry.Exception;
+            const int DefaultBufferSize = 1024;
+            var output = new ArrayBufferWriter<byte>(DefaultBufferSize);
+            using (var writer = new Utf8JsonWriter(output, FormatterOptions.JsonWriterOptions))
+            {
+                writer.WriteStartObject();
+
+
+                string timestamp = null;
+                var timestampFormat = FormatterOptions.TimestampFormat;
+                if (timestampFormat != null)
+                {
+                    var dateTime = FormatterOptions.UseUtcTimestamp ? DateTime.UtcNow : DateTime.Now;
+                    timestamp = dateTime.ToString(timestampFormat);
+                }
+                writer.WriteString("timestamp", timestamp);
+                writer.WriteNumber("eventId", eventId);
+                writer.WriteString("logLevel", GetLogLevelString(logLevel));
+                writer.WriteString("category", category);
+                writer.WriteString("message", message);
+
+                if (exception != null)
+                {
+                    writer.WriteStartObject("exception");
+                    writer.WriteString("message", exception.Message.ToString());
+                    writer.WriteString("type", exception.GetType().ToString());
+                    writer.WriteStartArray("stackTrace");
+                    if (exception?.StackTrace != null)
+                    {
+                        foreach (var stackTraceLines in exception?.StackTrace?.Split(Environment.NewLine))
+                        {
+                            JsonSerializer.Serialize<string>(writer, stackTraceLines);
```

Comment:
>JsonSerializer.Serialize<string>(writer, stackTraceLines); [](start = 28, length = 58)

I am wondering, does Json serializer cannot handle the line breaking?

do we have any version of the serializer can take a span instead so we can avoid the allocations?

Reply:
> does Json serializer cannot handle the line breaking?

No, the serializer wouldn't know to split on new lines and treat the result as a JSON array.

> do we have any version of the serializer can take a span instead so we can avoid the allocations of .Split call?

Yes, the serializer & writer both have span overloads, so when https://github.com/dotnet/runtime/issues/934 is implemented (or a similar API is used), we can avoid the allocations.

Reply:
We could switch to span overloads for split once it becomes available in the future

Reply:
I'm not convinced we need to split for the stack.

Reply:
So we need to decide on string (with new lines) vs string[] as @davidfowl suggests.  If we go with string[] we shouldn't be using serializer, we should be slicing out the lines and writing each slice, which I think is https://github.com/dotnet/runtime/blob/4cb924de165aa8e5373a109a68611d422aa72a09/src/libraries/System.Text.Json/ref/System.Text.Json.cs#L460.